src/web/core/src/main/java/org/geoserver/web/data/workspace/WorkspaceProvider.java

```diff
@@ -53,29 +86,121 @@ public WorkspaceProvider() {
+                Filter excludeDefault = Predicates.notEqual("id", defaultWorkspace.getId());
+                filter = Predicates.and(excludeDefault, filter);
+                // and add it to the head or tail as appropriate below
+            } else {
+                defaultWorkspace = null;
+            }
+        } else if (null != property) {
+            throw new IllegalStateException("Unknown sort property " + property);
+        }
+
+        LinkedList<WorkspaceInfo> list;
+        try (CloseableIterator<WorkspaceInfo> items =
+                catalog.list(WorkspaceInfo.class, filter, (int) first, (int) count, sortOrder)) {
+            Stream<WorkspaceInfo> stream = Streams.stream(items);
+            if (null != defaultWorkspace) {
+                WorkspaceInfo def = defaultWorkspace;
+                stream = stream.map(item -> decorateDefault(def, item));
+            }
+            list = stream.collect(Collectors.toCollection(LinkedList::new));
+        }
+        if (property == DEFAULT && defaultWorkspace != null) {
+            // and add it to the head or tail as appropriate
+            if (sort.isAscending()) {
+                list.addFirst(defaultWorkspace);
+            } else {
+                list.addLast(defaultWorkspace);
+            }
+        }
+        return list.iterator();
     }
 
     @Override
     protected List<Property<WorkspaceInfo>> getProperties() {
-        List<Property<WorkspaceInfo>> modifiedPropertiesList =
-                PROPERTIES.stream().map(c -> c).collect(Collectors.toList());
+        List<Property<WorkspaceInfo>> modifiedPropertiesList = new ArrayList<>(PROPERTIES);
         // check geoserver properties
-        if (GeoServerApplication.get()
-                .getGeoServer()
-                .getSettings()
-                .isShowCreatedTimeColumnsInAdminList())
+        SettingsInfo settings = getSettings();
+        if (settings.isShowCreatedTimeColumnsInAdminList())
             modifiedPropertiesList.add(CREATED_TIMESTAMP);
-        if (GeoServerApplication.get()
-                .getGeoServer()
-                .getSettings()
-                .isShowModifiedTimeColumnsInAdminList())
+        if (settings.isShowModifiedTimeColumnsInAdminList())
             modifiedPropertiesList.add(MODIFIED_TIMESTAMP);
         return modifiedPropertiesList;
     }
 
+    protected SettingsInfo getSettings() {
+        return GeoServerApplication.get().getGeoServer().getSettings();
+    }
+
     @Override
     protected IModel<WorkspaceInfo> newModel(WorkspaceInfo object) {
         return new WorkspaceDetachableModel(object);
     }
+
+    private void validate(Long first, Long count) {
+        if (first > Integer.MAX_VALUE
+                || first < Integer.MIN_VALUE
+                || count > Integer.MAX_VALUE
+                || count < Integer.MIN_VALUE) {
+            throw new IllegalArgumentException(); // TODO Possibly change catalog API to use long
+        }
```

Comment:
Saw the same on the LayerProvider. Well, we'll consider it when GeoServer can easily load 2 billion layers or workspaces. I think we'll need some hardware advancements, and possibly more code improvements :rofl:

Reply:
lol yeah this one started as a c&p of that one